### ISSUE TITLE
Creation of experiment results in TypeError (incorrect path)

### ISSUE THREAD

**Issue by proletarian-tear** [SAP/machine-learning-lab#6]
When trying to create an experiment with
`exp = env.create_experiment("Test experiment)`

It results in a:
`TypeError: expected str, bytes or os.PathLike object, not NoneType`

The expected behavior would be the creation of the exp object, and nothing printed out. Creation of an experiment was also tried in a _welcome_ notebook, with identical steps (inside the same project), and everything works as expected.

**Steps to reproduce the issue:**

This is done in a _hyperparameter-optimization-tutorial_ notebook, so at the point of execution, there are all of the necessary imports:
```
from __future__ import division, print_function
import logging, os, sys, time

logging.basicConfig(format='[%(levelname)s] %(message)s', level=logging.INFO, stream=sys.stdout)

import tqdm
tqdm.tqdm = tqdm.tqdm_notebook

%load_ext autoreload
%autoreload 2

import numpy as np
import pandas as pd
import fastText

from lab_client import Environment
from lab_client.commons import file_utils
```

And an initialized environment (project for which was created beforehand)
`env = Environment(project="first-test-project")`

**Technical details:**

- Host Machine OS (Windows/Linux/Mac): Windows 10
- Lab version :  `echo $SERVICE_VERSION` prints out a blank line, but the pulled from github version of lab is 0.1.0
- Docker version : 
```
 Client: Docker Engine - Community
 Azure integration  0.1.15
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:43:18 2020
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.12
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       48a66213fe
  Built:            Mon Jun 22 15:49:27 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```
- Host Machine OS (Windows/Linux/Mac): Windows 10
- Command used to start the ML Lab :` docker run --rm --env LAB_ACTION=install --env LAB_PORT=8080 --env BACKEND_SERVICE_IMAGE=docker.pkg.github.com/sap/machine-learning-lab/lab-service:0.1.0 --env MODEL_SERVICE_IMAGE=docker.pkg.github.com/sap/machine-learning-lab/lab-model-service:0.1.0 --env WORKSPACE_IMAGE=docker.pkg.github.com/sap/machine-learning-lab/ml-workspace-lab:0.1.0 -v /var/run/docker.sock:/var/run/docker.sock docker.pkg.github.com/sap/machine-learning-lab/lab-service:0.1.0`
- Browser (Chrome/Firefox/Safari): Firefox 81.0 (64-bit)

Traceback of the TypeError:
---------------------------------------------------------------------------
```
TypeError                                 Traceback (most recent call last)
<ipython-input-25-d80b4d5a2edd> in <module>
      1 # Initialize experiment
----> 2 exp = env.create_experiment("Test experiment")

/opt/conda/lib/python3.7/site-packages/lab_client/environment.py in create_experiment(self, name)
    433         The created #Experiment instance.
    434         """
--> 435         self.active_exp = Experiment(self, name, context_symbol_table=experiment_utils.get_caller_symbol_table())
    436         return self.active_exp

/opt/conda/lib/python3.7/site-packages/lab_client/handler/experiment_handler.py in __init__(self, env, name, auto_sync, track_file_events, upload_code_script, upload_code_repo, redirect_logs, context_symbol_table, root_folder, group_key, experiment_key)
    134 
    135         # Initialize experiment metadata
--> 136         self._exp_metadata = self._init_default_metadata()
    137 
    138         if group_key:

/opt/conda/lib/python3.7/site-packages/lab_client/handler/experiment_handler.py in _init_default_metadata(self)
    162         exp_metadata.host = experiment_utils.get_host_info()
    163         exp_metadata.git = experiment_utils.get_git_info(self._exp_script_dir)
--> 164         exp_metadata.dependencies = experiment_utils.get_python_dependencies(self._context_symbol_table)
    165         exp_metadata.resources = ExperimentResources()
    166 

/opt/conda/lib/python3.7/site-packages/lab_client/utils/experiment_utils.py in get_python_dependencies(symbol_table)
    103     # all dependencies with versions
    104     deps = []
--> 105     _, _, found_deps = dependency_utils.gather_sources_and_dependencies(symbol_table)
    106 
    107     if found_deps:

/opt/conda/lib/python3.7/site-packages/lab_client/commons/dependency_utils.py in gather_sources_and_dependencies(globs, base_dir, source_discovery_strategy, dependency_discovery_strategy)
    513 
    514     gather_sources = source_discovery_strategies[source_discovery_strategy]
--> 515     sources = gather_sources(globs, base_dir)
    516     if main is not None:
    517         sources.add(main)

/opt/conda/lib/python3.7/site-packages/lab_client/commons/dependency_utils.py in get_sources_from_imported_modules(globs, base_path)
    462 
    463 def get_sources_from_imported_modules(globs, base_path):
--> 464     return get_sources_from_modules(iterate_imported_modules(globs), base_path)
    465 
    466 

/opt/conda/lib/python3.7/site-packages/lab_client/commons/dependency_utils.py in get_sources_from_modules(module_iterator, base_path)
    431             continue
    432 
--> 433         filename = os.path.abspath(mod.__file__)
    434         if filename not in sources and \
    435                 is_local_source(filename, modname, base_path):

/opt/conda/lib/python3.7/posixpath.py in abspath(path)
    376 def abspath(path):
    377     """Return an absolute path."""
--> 378     path = os.fspath(path)
    379     if not isabs(path):
    380         if isinstance(path, bytes):

TypeError: expected str, bytes or os.PathLike object, not NoneType
```

**Comment by raethlein** [comment]
Hey @proletarian-tear, thanks for providing the details to your issue. Unfortunately, I cannot reproduce the bug. I am trying it on macOS, but since you don't mount any volumes besides the Docker socket, I would assume it should behave the same. One thing I observed is that there is an issue in the notebook's imports, as I cannot execute `import fastText` but `import fasttext`. So, executing the imports cell should already fail. Besides that, have you tried to restart the Jupyter kernel?

**Comment by JNKielmann** [comment]
Hi all,
I have looked a little bit into this issue and was able to reproduce it.

The problem is that in this line
https://github.com/SAP/machine-learning-lab/blob/master/libraries/ml-lab-py/lab_client/commons/dependency_utils.py#L430
It is check if the module object has the attribute __file__ but it is not checked if it is None.
According to:
https://docs.python.org/3/reference/import.html#__file__
> __file__ is optional. If set, this attribute’s value must be a string. 

The __file__ attribute always needs to be a string or is not set, so it could never be None. In that case the current implementation is fine. However, in python 3.7 the following change was introduced:
https://docs.python.org/3/whatsnew/3.7.html#other-cpython-implementation-changes
> Fixed some consistency problems with namespace package module attributes. Namespace module objects now have an __file__ that is set to None (previously unset) [...]

With this I was able to reproduce the issue:
1. I created a conda environment with python 3.7
2. I created a folder named "namespace" without an `__init__.py` file.
3. Next to the "namespace" folder I created a Jupyter notebook and executed the following code:

![Reproduced error](https://user-images.githubusercontent.com/10096239/96444086-ef15f300-120d-11eb-8667-d66d66288f7a.png)

This could also explain the error from above, if there was a folder called "fastText" next to the notebook, it would have been imported instead of the actual library which would have required `import fasttext`.

@raethlein If you want, I can create a PR for this and also take a look at the other places in the code where `hasattr(mod, '__file__')` is used
